File: lessons/04_03/Python/2_abstraction.py
class Animal:
  def __init__(self, health: int, speed: float, environment: str) -> None:
    self.__health = health
    self.__speed = speed
    self.__environment = environment

  @property
  def health(self) -> int:
    return self.__health

  @health.setter
  def health(self, health: int) -> None:
    self.__health = health

  @property
  def speed(self) -> float:
    return self.__speed

  @speed.setter
  def speed(self, speed: float) -> None:
    self.__speed = speed

  def animal_info(self) -> None:
    print(f"Animal lives in {self.__environment}")

animals = [
  ("Pigeon", Animal(100, 40, "Air")),
  ("Dog", Animal(90, 25, "Earth")),
  ("Dolphin", Animal(65, 70, "Water"))
]

for name, animal in animals:
  print(f"\t{name.upper()} INFO:")
  print(f"{name} health: {animal.health}")
  print(f"{name} speed: {animal.speed}")
  animal.animal_info()
  print()

File: lessons/04_03/C#/2_abstraction.cs
using System;

class Animal {
    private int health;
    private float speed;
    private string environment;

    public Animal(int health, float speed, string environment) {
        this.health = health;
        this.speed = speed;
        this.environment = environment;
    }

    public int GetHealth() {
        return health;
    }

    public void SetHealth(int health) {
        this.health = health;
    }

    public float GetSpeed() {
        return speed;
    }

    public void SetSpeed(float speed) {
        this.speed = speed;
    }

    public void AnimalInfo() {
        Console.WriteLine($"Animal lives in {environment}");
    }
}

class Program {
    static void Main() {
        (string name, Animal instance)[] animals = {
            ("Pigeon", new Animal(100, 40, "Air")),
            ("Dog", new Animal(90, 25, "Earth")),
            ("Dolphin", new Animal(65, 70, "Water"))
        };

        foreach (var (name, instance) in animals) {
            Console.WriteLine($"\t{name.ToUpper()} INFO:");
            Console.WriteLine($"{name} health: {instance.GetHealth()}");
            Console.WriteLine($"{name} speed: {instance.GetSpeed()}");
            instance.AnimalInfo();
            Console.WriteLine();
        }
    }
}
